middleware/plugin/middleware.go

```diff
@@ -0,0 +1,164 @@
+ */
+
+import (
+	"net/http"
+	"net/url"
+	"sync/atomic"
+	"time"
+
+	"github.com/Southclaws/fault"
+	"github.com/rs/zerolog/log"
+	genPlugin "github.com/siemens/wfx/generated/plugin"
+	"github.com/siemens/wfx/generated/plugin/client"
+	"github.com/siemens/wfx/middleware/logging"
+)
+
+type MW struct {
+	plugin    Plugin
+	chMessage chan Message
+	quitChan  chan error
+}
+
+func NewMiddleware(plugin Plugin, quitChan chan error) (*MW, error) {
+	contextLogger := log.With().Str("plugin", plugin.Name()).Logger()
+	contextLogger.Debug().Msg("Creating new plugin middleware")
+	chMessage, err := plugin.Start(quitChan)
+	if err != nil {
+		_ = plugin.Stop()
+		return nil, fault.Wrap(err)
+	}
+	return &MW{
+		plugin:    plugin,
+		chMessage: chMessage,
+		quitChan:  quitChan,
+	}, nil
+}
+
+func (mw *MW) Wrap(next http.Handler) http.Handler {
+	var cookieCounter atomic.Uint64
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log := logging.LoggerFromCtx(r.Context()).With().Str("plugin", mw.plugin.Name()).Logger()
+
+		req := convertRequest(r, cookieCounter.Add(1))
+		msg := Message{
+			request:  req,
+			response: make(chan genPlugin.PluginResponseT, 1),
+		}
+
+		log.Debug().Msg("Sending request to plugin")
+		start := time.Now()
+		mw.chMessage <- msg
+		log.Debug().Msg("Waiting for plugin response")
+		resp := <-msg.response
+		duration := time.Since(start)
+		log.Info().Dur("duration", duration).Msg("Received plugin response")
+
+		if resp.Payload != nil {
+			switch resp.Payload.Type {
+			case genPlugin.Payloadgenerated_plugin_client_Response:
+				log.Info().Msg("Sending client response provided by plugin")
+				val := resp.Payload.Value.(*client.ResponseT)
+				for _, h := range val.Envelope {
+					for _, value := range h.Values {
+						w.Header().Add(h.Name, value)
+					}
+				}
+				switch val.Status {
+				case client.ResponseStatusAccept:
+					w.WriteHeader(http.StatusOK)
+				case client.ResponseStatusModified:
+					w.WriteHeader(http.StatusOK)
+				case client.ResponseStatusDeny:
+					w.WriteHeader(http.StatusForbidden)
+				}
+				_, _ = w.Write(val.Content)
+				return
+			case genPlugin.Payloadgenerated_plugin_client_Request:
+				log.Info().Msg("Request was modified by plugin")
+				// override http.Request with the response
+				val := resp.Payload.Value.(*client.RequestT)
+
+				if parsedURL, err := url.Parse(val.Destination); err != nil {
+					log.Err(err).Str("destination", val.Destination).Msg("Failed to parse destination")
+				} else {
+					r.URL = parsedURL
+				}
+
+				// delete existing headers
+				for k := range r.Header {
+					delete(r.Header, k)
+				}
+				if len(val.Envelope) > 0 {
+					if r.Header == nil {
+						r.Header = make(http.Header)
+					}
+					for _, h := range val.Envelope {
+						for _, value := range h.Values {
+							r.Header.Add(h.Name, value)
+						}
+					}
+				}
+			default:
+				log.Warn().Int("type", int(resp.Payload.Type)).Msg("Unhandled payload type")
```

Comment:
Shouldn't we terminate in this case? The plugin is not playing along our interface rules...